utils/s3/credentials_providers/sts_assume_role_credentials_provider.cc

```diff
@@ -0,0 +1,98 @@
+    if (!is_time_to_refresh() || !creds) {
+        co_return co_await update_credentials();
+    }
+}
+
+future<> sts_assume_role_credentials_provider::update_credentials() {
+    auto req = http::request::make("POST", sts_host, "/");
+    // Just set this version
+    // https://github.com/aws/aws-sdk-cpp/blob/8d68be52dcad85095753e069a4355e241f1edb1c/generated/src/aws-cpp-sdk-sts/source/model/AssumeRoleRequest.cpp#L143
+    req.query_parameters["Version"] = "2011-06-15";
+    req.query_parameters["DurationSeconds"] = format("{}", session_duration);
+    req.query_parameters["Action"] = "AssumeRole";
+    req.query_parameters["RoleSessionName"] = format("{}", utils::make_random_uuid());
+    req.query_parameters["RoleArn"] = role_arn;
+    auto factory = std::make_unique<utils::http::dns_connection_factory>(sts_host, port, is_secured, sts_logger);
+    http::experimental::client http_client(std::move(factory), 1, http::experimental::client::retry_requests::yes);
+    co_return co_await http_client.make_request(
+        std::move(req),
+        [this](const http::reply&, input_stream<char>&& in) -> future<> {
+            auto input = std::move(in);
+            auto body = co_await util::read_entire_stream_contiguous(input);
+            parse_creds(body);
+        },
+        http::reply::status_type::ok);
+}
+
+void sts_assume_role_credentials_provider::parse_creds(sstring& body) {
+    auto get_node_safe = []<typename T>(const T* node, const std::string_view node_name) {
+        auto child = node->first_node(node_name.data());
+        if (!child) {
+            throw std::runtime_error(seastar::format("'{}' node is missing in AssumeRole response", node_name));
+        }
+        return child;
+    };
+    auto doc = std::make_unique<rapidxml::xml_document<>>();
+    try {
+        doc->parse<0>(body.data());
+    } catch (const rapidxml::parse_error& e) {
+        throw std::runtime_error("Cannot parse AssumeRole response");
+    }
+
+    const auto root = get_node_safe(doc.get(), "AssumeRoleResponse");
+    const auto result = get_node_safe(root, "AssumeRoleResult");
+    const auto assumed_role_user = get_node_safe(result, "AssumedRoleUser");
+    sts_logger.info("Assuming role for ARN: {}, Assumed Role ID: {}",
+                    assumed_role_user->first_node("Arn")->value(),
+                    assumed_role_user->first_node("AssumedRoleId")->value());
+
+    // Access the Credentials node
+    const auto credentials = get_node_safe(result, "Credentials");
+    creds.access_key_id = get_node_safe(credentials, "AccessKeyId")->value();
+    creds.secret_access_key = get_node_safe(credentials, "SecretAccessKey")->value();
+    creds.session_token = get_node_safe(credentials, "SessionToken")->value();
+    // Set the expiration to one minute earlier to ensure credentials are renewed slightly before they expire
+    creds.expires_at = std::chrono::system_clock::now() + std::chrono::seconds(session_duration - 60);
```

Comment:
It doesn't look correct to update `creds` in place. In case get_node_safe() throws in the middle, the `creds` object happens to be half-set with new values and half-set with old ones

Reply:
Then following request would fail with expired creds message, this scenario will be handled in next PR after [this one](https://github.com/scylladb/scylladb/pull/21933) will be approved. Are we ok with such a behavior?

Reply:
> Then following request would fail with expired creds message

Or with invalid creds message, we don't know exactly how intermixed creds would look like and what would aws think of it :(

Reply:
ok, instead of guessing what will happen in such a case I will move the creds update outside